README.md
# Starting with Clojure

A Clojure library designed to ... well, that part is up to you.

## Usage

Understanding how the language works and the getting the basic commands.

## License

Copyright © 2023 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-06-20
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-06-20
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/estoque/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/estoque/compare/0.1.0...0.1.1


src/estoque/aula2.clj
(ns estoque.aula2)

(defn imprime-mensagem []
  (println "------------------------")
  (println "Bem vindo(a) ao estoque!"))

(imprime-mensagem)



; * valor-bruto 0.9

(defn aplica-desconto [valor-bruto]
  (* valor-bruto 0.9))

(println (aplica-desconto 100))


(defn valor-descontado
  "Retorna o valor descontado que é 90% do valor bruto"
  [valor-bruto]
  (* valor-bruto 0.9))

(println
  (valor-descontado 100))


(defn valor-descontado [valor-bruto] (* valor-bruto 0.9))


(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (* valor-bruto (- 1 0.10)))

(valor-descontado 100)




(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (def desconto 0.10)
  (* valor-bruto (- 1 desconto)))

(valor-descontado 100)




(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (let [desconto 0.10]
    (* valor-bruto (- 1 desconto))
    )
  )

(valor-descontado 100)











(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (let [desconto 0.10]
    (println "Calculando desconto de " desconto)
    (* valor-bruto (- 1 desconto))
    )
  )

(valor-descontado 100)









(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (
    let [desconto 0.10]
    (
      println "Calculando desconto de " desconto)
    (* valor-bruto (- 1 desconto))
    )
  )

(valor-descontado 100)






(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (let [desconto 0.10]
    (println "Calculando desconto de " desconto)
    (* valor-bruto (- 1 desconto))))

(valor-descontado 100)




(class 90.0)



(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (let [desconto (/ 10 100)]
    (println "Calculando desconto de " desconto)
    (* valor-bruto (- 1 desconto))))

(valor-descontado 100)



(class 90N)
(class 90M)








(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (let [taxa-de-desconto (/ 10 100)
        desconto         (* valor-bruto taxa-de-desconto)]
    (println "Calculando desconto de " desconto)
    (- valor-bruto desconto)))

(valor-descontado 100)





(> 500 100)
(< 500 100)

(if (> 500 100)
  (println "maior")
  (println "menor ou igual"))

(if (> 50 100)
  (println "maior")
  (println "menor ou igual"))









(defn valor-descontado
  "Retorna o valor com desconto de 10% se o valor bruto for estritamente maior que 100."
  [valor-bruto]
  (if (> valor-bruto 100)
    (let [taxa-de-desconto (/ 10 100)
          desconto         (* valor-bruto taxa-de-desconto)]
      (println "Calculando desconto de " desconto)
      (- valor-bruto desconto))))

(valor-descontado 1000)

(valor-descontado 1000)






(defn valor-descontado
  "Retorna o valor com desconto de 10% se o valor bruto for estritamente maior que 100."
  [valor-bruto]
  (if (> valor-bruto 100)
    (let [taxa-de-desconto (/ 10 100)
          desconto         (* valor-bruto taxa-de-desconto)]
      (println "Calculando desconto de " desconto)
      (- valor-bruto desconto))
    valor-bruto))

(valor-descontado 1000)

(valor-descontado 100)



src/estoque/aula3.clj
(ns estoque.aula3)

(defn valor-descontado
  "Retorna o valor com desconto de 10% se o valor bruto for estritamente maior que 100."
  [valor-bruto]
  (if (> valor-bruto 100)
    (let [taxa-de-desconto (/ 10 100)
          desconto         (* valor-bruto taxa-de-desconto)]
      (- valor-bruto desconto))
    valor-bruto))

(println (valor-descontado 1000))

(println (valor-descontado 100))

; PREDICATE
(defn aplica-desconto?
  [valor-bruto]
  (if (> valor-bruto 100)
    true
    false))

(println (aplica-desconto? 1000))
(println (aplica-desconto? 100))

(defn valor-descontado
  "Retorna o valor com desconto de 10% se o valor bruto for estritamente maior que 100."
  [valor-bruto]
  (if (aplica-desconto? valor-bruto)
    (let [taxa-de-desconto (/ 10 100)
          desconto         (* valor-bruto taxa-de-desconto)]
      (- valor-bruto desconto))
    valor-bruto))

(println (valor-descontado 1000))
(println (valor-descontado 100))





(defn aplica-desconto?
  [valor-bruto]
  (println "chamando nova versão")
  (if (> valor-bruto 100)
    true))

(println (valor-descontado 1000))
(println (valor-descontado 100))







(defn aplica-desconto?
  [valor-bruto]
  (println "chamando nova versão when")
  (when (> valor-bruto 100)
    true))

(println (valor-descontado 1000))
(println (valor-descontado 100))





(defn aplica-desconto?
  [valor-bruto]
  (println "chamando nova versão direta")
  (> valor-bruto 100))

(println (valor-descontado 1000))
(println (valor-descontado 100))





(println "#######################################")

(defn mais-caro-que-100?
  [valor-bruto]
  (> valor-bruto 100))


(defn valor-descontado
  "Retorna o valor com desconto de 10% se deve aplicar desconto."
  [aplica? valor-bruto]
  (if (aplica? valor-bruto)
    (let [taxa-de-desconto (/ 10 100)
          desconto         (* valor-bruto taxa-de-desconto)]
      (- valor-bruto desconto))
    valor-bruto))

(println (valor-descontado mais-caro-que-100? 1000))
(println (valor-descontado mais-caro-que-100? 100))

; HIGHER ORDER FUNCTIONS
; IMUTABILIDADE (a exemplos dos vetores)


(println "#### Função sem nome (anônima) ###########################")
(println (valor-descontado (fn [valor-bruto] (> valor-bruto 100)) 1000))
(println (valor-descontado (fn [valor-bruto] (> valor-bruto 100)) 100))


(println "#### Função sem nome (anônima) abreviada ###########################")
(println (valor-descontado #(> %1 100) 1000))
(println (valor-descontado #(> %1 100) 100))


(println "#### Função sem nome (anônima) abreviada só com um parâmetro ###########################")
(println (valor-descontado #(> % 100) 1000))
(println (valor-descontado #(> % 100) 100))

(def mais-caro-que-100? #(> % 100))
(println (valor-descontado mais-caro-que-100? 1000))
(println (valor-descontado mais-caro-que-100? 100))




src/estoque/aula4.clj
(ns estoque.aula4)

(def precos [30 700 1000])

(println (precos 0))
(println (get precos 0))
(println "valor padrão nulo:" (get precos 17))
(println "valor padrão 0:" (get precos 17 0))


(println (conj precos 5))
(println precos)


(println (+ 5 1))
(println (inc 5))
(println (update precos 0 inc))
(println (update precos 2 inc))


(defn soma-3
  [valor]
  (println "Estou somando um em"valor)
  (+ valor 3))

(println (update precos 0 soma-3))






(defn aplica-desconto?
  [valor-bruto]
  (> valor-bruto 100))

(defn valor-descontado
  "Retorna o valor com desconto de 10% se o valor bruto for estritamente maior que 100."
  [valor-bruto]
  (if (aplica-desconto? valor-bruto)
    (let [taxa-de-desconto (/ 10 100)
          desconto         (* valor-bruto taxa-de-desconto)]
      (- valor-bruto desconto))
    valor-bruto))

(println (map valor-descontado precos))

(println (range 10))
(println (filter even? (range 10)))




(println (filter aplica-desconto? precos))
(println (map valor-descontado (filter aplica-desconto? precos)))

(println (reduce + precos))

(defn minha-soma
  [valor-1 valor-2]
  (println "Somando"valor-1"e"valor-2)
  (+ valor-1 valor-2))

(println "######")
(println (reduce minha-soma precos))
(println (reduce minha-soma (range 10)))
(println (reduce minha-soma 0 [15]))










src/estoque/aula5.clj
(ns estoque.aula5)

(def estoque {"Mochila" 5, "Camisetas" 10})
(println estoque)


(def estoque
  {"Mochila" 5
   "Camiseta" 10})

(println estoque)

(println "Temos" (count estoque) "elementos")
(println "As chaves são:" (keys estoque))
(println "Os valores são:" (vals estoque))

(println "##########")

(def estoque
  {:mochila 10
   :camiseta 5})

(println "Temos" (count estoque) "elementos")
(println "As chaves são:" (keys estoque))
(println "Os valores são:" (vals estoque))

(println (assoc estoque :cadeira 3))
(println estoque)

(println estoque)
(println (update estoque :mochila inc))
(println estoque)

(defn tira-um
  [valor]
  (println "Tirando um de"valor)
  (- valor 1))

(println (update estoque :mochila #(- % 1)))

(println (dissoc estoque :mochila))






(println "################")

(def pedido
  {:mochila {:quantidade 2, :preco 80}
   :camiseta {:quantidade 3, :preco 40}})
(println pedido)


(def pedido (assoc pedido :chaveiro {:quantidade 1, :preco 10}))
(println pedido)
(println (pedido :mochile))                                 ;mapa como função
(println (get pedido :caneta {}))                           ;mapa com valor padrão
(println (:mochila pedido))
(println (:cadeira pedido))
(println (:cadeira pedido {}))

(println (:quantidade (:mochila pedido)))
(println (:quantidade (:mochila (update-in pedido [:mochila :quantidade] inc))))





; THREADING FIRST
(println (-> pedido
             :mochila
             :quantidade))

(-> pedido
    :mochila
    :quantidade
    println)















src/estoque/aula6.clj
(ns estoque.aula6)

(def pedido
  {:mochila {:quantidade 2, :preco 80}
   :camiseta {:quantidade 3, :preco 40}})

(defn imprime-e-15
  [valor]
  (println "valor" (class valor) valor)
  15)

(println (map imprime-e-15 pedido))


;(defn imprime-e-15
;  [chave valor]
;  (println "valor" valor "e" chave)
;  15)
;
;(println (map imprime-e-15 pedido))

(println "######")
(defn imprime-e-15
  [[chave valor]]
  (println "valor" valor "e chave" chave)
  15)

(println (map imprime-e-15 pedido))

(defn preco-dos-produto
  [[chave valor]]
  (* (:quantidade valor) (:preco valor)))

(println (map preco-dos-produto pedido))
(println (reduce + (map preco-dos-produto pedido)))




(println "Versão com parâmetro da 'chave' que não era utilziado substituído por um _")
(defn preco-dos-produto
  [[_ valor]]
  (* (:quantidade valor) (:preco valor)))

(println "Total por produto:" (map preco-dos-produto pedido))
(println "Total do pedido:" (reduce + (map preco-dos-produto pedido)))

(defn total-do-pedido
  [pedido]
  (reduce + (map preco-dos-produto pedido)))

(println (total-do-pedido pedido))


; THRED LAST: o parâmetro passado pra função vai no final da chamada.
; Então o (map preco-dos-produto ,,,) vai ter o parâmetro onde tem as vírgulas que eu adicionei;
; já no thread first, o parâmetro vem pro primeiro: assim -> (map ,,, preco-dos-produto)
(defn total-do-pedido
  [pedido]
  (->> pedido
      (map preco-dos-produto)
      (reduce +)))

(println "Total do pedido" (total-do-pedido pedido))





(defn preco-total-do-prduto
  [produto]
  (println "Produto"produto)
  (* (:quantidade produto) (:preco produto)))

(defn preco-total-do-pedido
  [pedido]
  (->> pedido
       vals
       (map preco-total-do-prduto)
       (reduce +)))

(println "Total do pedido" (preco-total-do-pedido pedido))






(def pedido
  {:mochila {:quantidade 2, :preco 80}
   :camiseta {:quantidade 3, :preco 40}
   :chaveiro {:quantidade 1}})

(defn gratuito?
  [[_ item]]
  (<= (get item :preco 0) 0))

(println (filter gratuito? pedido))



(defn gratuito?
  [item]
  (<= (get item :preco 0) 0))

(println (filter (fn [[chave item]] (gratuito? item)) pedido))


(defn gratuito?
  [item]
  (<= (get item :preco 0) 0))

(println (filter #(gratuito? (second %)) pedido))




(defn pago?
  [item]
  (not (gratuito? item)))

(println (pago? {:preco 0}))
(println (pago? {:preco 30}))


(println ((comp not gratuito?) {:preco 50}))
(println ((comp not gratuito?) {:preco 0}))

(def pago? (comp not gratuito?))
(println (pago? {:preco 50}))
(println (pago? {:preco 0}))






(def clientes
  [{:nome "Guilherme" :certificados ["Clojure" "Java" "Machine Learning"]}
   {:nome "Paulo" :certificados ["Java" "Ciência da Computação"]}
   {:nome "Daniela" :certificados ["Arquitetura" "Gastronomia"]}])

(println (->> clientes
              (map :certificados)
              (map count)
              (reduce +)))














